### ISSUE TITLE
launchsettings ignored by docker tools

### ISSUE THREAD

**Issue by machie27** [microsoft/DockerTools#215]
I recently upgraded to .net core 3.0.
As I upgraded and need some specific ports, see below:
```

  "Docker Test": {
      "commandName": "Docker",
      "launchBrowser": true,
      "launchUrl": "{Scheme}://{ServiceHost}:{ServicePort}/page",
      "environmentVariables": {
        "ASPNETCORE_Kestrel__Certificates__Default__Path": "/https/aspnetapp.pfx",
        "ASPNETCORE_URLS": "https://+:44374;http://+:5000",
        "ASPNETCORE_Kestrel__Certificates__Default__Password": "****",
        "ASPNETCORE_HTTPS_PORT": "44374",
        "ASPNETCORE_ENVIRONMENT": "Test"
      },
      "httpPort": 5000,
      "useSSL": true,
      "sslPort": 44374
    },

```
However the following is begin deployed:
```
>docker run -dt -v "C:\Users\***\vsdbg\vs2017u5:/remote_debugger:rw" 
-v "C:\Users\***\source\repos\backend\***\***:/app" 
-v "C:\Users\***\source\repos\backend\***:/src" 
-v "C:\Users\***\AppData\Roaming\Microsoft\UserSecrets:/root/.microsoft/usersecrets:ro" -v "C:\Users\***\AppData\Roaming\ASP.NET\Https:/root/.aspnet/https:ro" 
-v "C:\Users\***\.nuget\packages\:/root/.nuget/fallbackpackages2" 
-v "C:\Program Files\dotnet\sdk\NuGetFallbackFolder:/root/.nuget/fallbackpackages" 
-e "DOTNET_USE_POLLING_FILE_WATCHER=1" 
-e "ASPNETCORE_ENVIRONMENT=Development" 
-e "ASPNETCORE_URLS=https://+:443;http://+:80" 
-e "ASPNETCORE_HTTPS_PORT=44374" 
-e "NUGET_PACKAGES=/root/.nuget/fallbackpackages2" 
-e "NUGET_FALLBACK_PACKAGES=/root/.nuget/fallbackpackages;/root/.nuget/fallbackpackages2" 
-p 5000:80 
-p 44374:443 
-v "C:\Users\***\.aspnet\https:/https/" --entrypoint tail  ***i:dev -f /dev/null
```
As you can see above settings get ignored as the ports are mapped 5000:80 and 44374:443 instead of 5000:5000 and 44374:44374. Also the environment is being set to Development instead of Test. This is not how it's expected to work.
below is the dockerfile as reference:
```
FROM mcr.microsoft.com/dotnet/core/aspnet:3.0.0-alpine3.9 AS base
WORKDIR /app
EXPOSE 5000 44374
FROM mcr.microsoft.com/dotnet/core/sdk:3.0.100-alpine3.9 AS build
WORKDIR /src 
COPY ["**.csproj", "Dop.Rapp.Common/"]
RUN dotnet restore "***.csproj"
COPY . .
WORKDIR "/src/*i"
RUN dotnet build "**.csproj" -c Release -o /app
FROM build AS publish
RUN dotnet publish "**.csproj" -c Release -o /app
FROM base AS final
WORKDIR /app
COPY --from=publish /app .
ENTRYPOINT ["dotnet", "**.dll"]
```

**Comment by bwateratmsft** [comment]
Thanks @machie27. @haniamr is investigating this and other issues around ports; we're trying to find ways to reduce port conflict failures while ideally also giving users more control.

**Comment by machie27** [comment]
Thanks Stephen. I’m glad you like the new Containers window!

For the issue above, we are preparing a fix that will be available in VS 16.5, in the meantime there’s a workaround that can help unblock you (note: only works in 16.4 preview 3 and above):

Modify your launchSettings.json to the following:
```
"Docker": {
      "commandName": "Docker",
      "launchBrowser": true,
      "launchUrl": "{Scheme}://{ServiceHost}:44374", #<== Set a fixed port
      "environmentVariables": {
        "ASPNETCORE_URLS": "https://+:44374;https://+:5000",
        "ASPNETCORE_HTTPS_PORT": "44374"
      },
      "publishAllPorts": true, #<== This is equivalent to the -P flag in 'docker run'
      "useSSL": true
    }
```
Add the following property to your .csproj file to override the port mappings:
`<DockerfileRunArguments>-p 44374:44374 -p 5000:5000</DockerfileRunArguments>`
What this does is that it will pass the flag “-P” to the “docker run” command, however you override the port mappings with the provided property above. Please let me know how this works for you and I’ll update this issue as soon as the fix is integrated.

Thanks,

-Hani
